python/ray/data/datasource/mongo_datasource.py

```diff
@@ -0,0 +1,185 @@
+        if len(pipeline) == 0 or "$match" not in pipeline[0]:
+            return {}
+        return pipeline[0]["$match"]
+
+    def get_read_tasks(self, parallelism: int) -> List[ReadTask]:
+        from bson.objectid import ObjectId
+
+        coll = self._client[self._database][self._collection]
+        match_query = self._get_match_query(self._pipeline)
+        partitions_ids = list(
+            coll.aggregate(
+                [
+                    {"$match": match_query},
+                    {"$bucketAuto": {"groupBy": "$_id", "buckets": parallelism}},
+                ],
+                allowDiskUse=True,
+            )
+        )
+
+        def make_block(
+            uri: str,
+            database: str,
+            collection: str,
+            pipeline: List[Dict],
+            min_id: ObjectId,
+            max_id: ObjectId,
+            right_closed: bool,
+            schema: Schema,
+            kwargs: dict,
+        ) -> Block:
+            from pymongoarrow.api import aggregate_arrow_all
+
+            # A range query over the partition.
+            match = [
+                {
+                    "$match": {
+                        "_id": {
+                            "$gte": min_id,
+                            "$lte" if right_closed else "$lt": max_id,
+                        }
+                    }
+                }
+            ]
+            client = pymongo.MongoClient(uri)
+            return aggregate_arrow_all(
+                client[database][collection], match + pipeline, schema=schema, **kwargs
+            )
+
+        read_tasks: List[ReadTask] = []
+
+        for i, partition in enumerate(partitions_ids):
+            metadata = BlockMetadata(
+                num_rows=partition["count"],
+                size_bytes=None,
+                schema=None,
+                input_files=None,
+                exec_stats=None,
+            )
+            min_id = partition["_id"]["min"]
+            max_id = partition["_id"]["max"]
+            right_closed = i == len(partitions_ids) - 1
+            read_task = ReadTask(
+                lambda uri=self._uri, database=self._database, collection=self._collection, pipeline=self._pipeline, min_id=min_id, max_id=max_id, right_closed=right_closed, schema=self._schema, kwargs=self._mongo_args: [  # noqa: E501
+                    make_block(
+                        uri,
+                        database,
+                        collection,
+                        pipeline,
+                        min_id,
+                        max_id,
+                        right_closed,
+                        schema,
+                        kwargs,
+                    )
+                ],
+                metadata,
+            )
+            read_tasks.append(read_task)
+
+        return read_tasks
+
+
+def _validate_database_collection_exist(client, database: str, collection: str):
+    db_names = client.list_database_names()
```

Comment:
why not just use `client.get_database`, `get_collection` APIs - https://api.mongodb.com/python/2.9.4/api/pymongo/mongo_client.html#pymongo.mongo_client.MongoClient.get_database ? We don't need to fetch all all database and collection names here.

Reply:
It's because this is just about checking existence, not creating database ("Useful for *creating* a Database..." in get_database).